Added a confirm command.

aRkedos opened the pull request:
#### ISSUE TYPE

- Improvement/feature implementation

#### RUNTIME ENVIRONMENT
<!-- Details of your runtime environment -->
<!-- Retrieve Python/ranger version and locale with `ranger -\-version` -->
- Operating system and version: Manjaro Linux 5.3.12-1 
- Terminal emulator and version: Termite v15
- Python version: 3.7.4
- Ranger version/commit: Ranger 1.9.2 4ff19f6
- Locale: de_DE.UTF8

#### CHECKLIST

- [X] The `CONTRIBUTING` document has been read **[REQUIRED]**
- [X] All changes follow the code style **[REQUIRED]**
- [X] All new and existing tests pass **[REQUIRED]**
- [ ] Changes require config files to be updated
    - [ ] Config files have been updated
- [ ] Changes require documentation to be updated
    - [ ] Documentation has been updated
- [ ] Changes require tests to be updated
    - [ ] Tests have been updated

#### DESCRIPTION
This PR adds a Command which can be used for any Ranger command and asks the user for confirmation if the command shall be executed.


#### MOTIVATION AND CONTEXT
User requested to possibilty to be asked for confirmation before any command in Issue #1620


#### TESTING
Make test has been run.
So far I tested the following commands:
- delete (if confirmation is a activated ranger correctly asks for conformation twice) 
- cd
- set
- chain with different commands. Worked, although chaining delete with cd can be risky.

aRkedos committed:
Added a confirm action

aRkedos commented on ranger/config/commands.py:
```diff
@@ -738,8 +738,8 @@ def is_directory_with_files(path):
-        confirm = self.fm.settings.confirm_on_delete
-        if confirm != 'never' and (confirm != 'multiple' or many_files):
+        confirm_setting = self.fm.settings.confirm_on_delete
+        if confirm_setting != 'never' and (confirm_setting != 'multiple' or many_files):
```
These changes were necessary due to potential namespace conflict since the new class is also named "confirm".

toonn commented on ranger/config/commands.py:
```diff
@@ -1959,3 +1959,25 @@ def tab(self, tabnum):
+    def execute(self):
+
+        from functools import partial
+        command = self.arg(1)
```
I don't like this as much. We should show the entire command *at least*. Ideally the command with macros resolved.

aRkedos replied:
Okay two questions here:
1. What do you mean by macros? I have done some searching through the codebse and on the Wiki here on Github, are macros just key bindings of commands? So if this is correct, you want to show which command exactly gets executed?

2. Shall I also show selection of files if you operate on commands that use a selection of files (e.g: delete or open_with)?

toonn replied:
Yes, selection's a good idea too. Though I think you should only show that if there's no arguments specified. In other words, always show just exactly what files it'll operate on.

toonn reviewed (changes requested):
One issue, the rest looks good to me.

aRkedos committed:
Show the complete command

aRkedos committed:
Reworked function with recursion for chaining

aRkedos committed:
Refactor the ask part into a helper function

aRkedos commented:
Okay this is the version we discussed in IRC. It works. Chained commands will be split up and called recursively.

toonn commented on ranger/config/commands.py:
```diff
@@ -1959,3 +1959,46 @@ def tab(self, tabnum):
+                ('n', 'N', 'y', 'Y'),
+            )
+        else:
+            arguments = self.rest(2).split(';')
```
Let's use a list comprehension here instead.

aRkedos replied:
I already tried that, I did something like `arguments = [arg.strip(';') for arg in self.args]`. But this separates arguments of a command from the command itsself, e.g. If you do `:confirm chain delete; echo hi` you would geld a list with `['confirm', 'chain', 'delete', 'echo', 'hi']`. 
So what benefit would a list comprehension have?

Please don't get me wrong I will try my best to make this PR fit the needs of ranger, just curious about the reasons in order to learn something :)

aRkedos replied:
Okay figured it out. Just had to iterate over self.rest(2).split(';') in the comprehension. Sometimes I fail to see the obvious. 
:)

toonn commented on ranger/config/commands.py:
```diff
@@ -1959,3 +1959,46 @@ def tab(self, tabnum):
+
+    def _chaining(self, recursion_count, commands):
+        from functools import partial
+        if recursion_count >= 0:
```
Why do we need this count?

aRkedos replied:
Well I could work with len() of the list, or delete a command from the list after every recursion and check if the len() is still greater 0

toonn replied:
Aren't you already splitting the list on `;`s?

aRkedos replied:
Yes but each command is one element of a list . I used the counter to determine how long the recursion should go on. When the counter reaches zero, the first element of the commands list will be called (commands will be called from last to first, I could reverse that order if preferred), on the next iteration the counter will be -1 and the recursion stops. To me it seemed pretty straight forward.

toonn commented on ranger/config/commands.py:
```diff
@@ -1959,3 +1959,46 @@ def tab(self, tabnum):
+        from functools import partial
+        if recursion_count >= 0:
+            command = commands[recursion_count]
+            self.fm.ui.console.ask(
```
Why is this duplicated here?

aRkedos replied:
You mean the module import? Or do you mean I could use the commands[recursion_count] directly in the message?

Well without the module import the partial function is not part of the namespace in the chaining function and would not be executed.

toonn replied:
I mean the entire ask part.

aRkedos replied:
Well if I do it recursively I would have to call on that function again and again I guess?

toonn replied:
What does recursion have to do with duplication of code?

aRkedos replied:
Nothing directly, but I need to call the ask part in the non-recursive part and in the recursive part. Well I could pack that into another function which gets called where needed.

aRkedos replied:
Okay packed it into a helper function and thus avoiding duplicated code and the duplicated module import.

toonn replied:
Can't you just stow it in the non-recursive part? And then "recurse" (to a depth of 1) even for the simple case?

toonn replied:
A helper might be fine too. Go with whatever seems clearest.

toonn reviewed (changes requested):
Starting to look better. Not quite happy yet though.

aRkedos committed:
Refined command build-up

Avoid extra whitespace on the last argument.

toonn committed:
Refactor to use more idiomatic python

`' '.join(list)` is a very useful and common idiom to join elements of a
list without adding an extra space. In this case `self.rest(1)` gives us
pretty much the same result though.

The `_chaining` recursion was equivalent to a loop over the arguments.

The `if` branches interpreted no arguments as `chain`.

Vifon commented on ranger/config/commands.py:
```diff
@@ -1990,3 +1990,34 @@ def make_safe_path(dst):
+            self._ask_for_confirmation(self.rest(1))
+
+    def _question_callback(self, command, answer):
+        if answer in ('y', 'Y'):
```
Let's use `answer.lower()` here, this way there will be no need to enumerate every variant by hand.

Vifon commented on ranger/config/commands.py:
```diff
@@ -1990,3 +1990,34 @@ def make_safe_path(dst):
+            self.fm.notify("Nothing to confirm.")
+        elif self.arg(1) == 'chain':
+            # TODO: This doesn't work for chained chain commands.
+            for command in [arg.lstrip() for arg in self.rest(2).split(';')]:
```
Will this ask for each `:chain` subcommand separately? I think it would make more sense to ask once for the whole chain. What was the rationale behind this decision?

toonn replied:
I think this stems from a misunderstanding of something I said. If we do listing of the files a command operates on then we'd have to recurse for chain. But the PR doesn't do that yet so all it would do is make the commands slightly more readable (because they're shorter) when confirming at the cost of having to confirm each command in a chain. But in addition the way the chain is split doesn't work, cf. the TODO I added.

toonn commented on ranger/config/commands.py:
```diff
@@ -1959,3 +1959,55 @@ def tab(self, tabnum):
+class confirm(Command):
+    """:confirm <command_to_be_confirmed> + <potential command arguments>
+
+        Can be used with any command. Ranger will ask if it should execute the
```
I could've sworn I caught this.

toonn committed:
Remove special case for chain

Convert to lower case rather than checking for membership of an
enumeration.
